## The Validation Set Approach



In [2]:
library(ISLR2)

set.seed(1)

dim(Auto)

[1] 392   9

In [3]:
# Select a random subset of 196 observations out of the 392 observations
train <- sample(392, 196)

In [5]:
# Use the subset option and run a linear regression on the Auto data 
lm_fit <- lm(mpg ~ horsepower, data = Auto, subset = train)

# Estimate the response for all 392 observations and use the mean function
# to calculate the MSE of 196 observations in the validation set 
attach(Auto)

mean((mpg - predict(lm_fit, Auto))[-train]^2)

[1] 23.26601

In [6]:
# We can use the poly() function to estimate the 
# test error for the quadratic regressions.
lm_fit2 <- lm(mpg ~ poly(horsepower, 2), data = Auto, subset = train)

attach(Auto)
mean((mpg - predict(lm_fit2, Auto))[-train]^2)

The following objects are masked from Auto (pos = 3):

    acceleration, cylinders, displacement, horsepower, mpg, name,
    origin, weight, year



[1] 18.71646

In [7]:
# We can use the poly() function to estimate the 
# test error for the cubic regressions.
lm_fit3 <- lm(mpg ~ poly(horsepower, 3), data = Auto, subset = train)

attach(Auto)
mean((mpg - predict(lm_fit3, Auto))[-train]^2)

The following objects are masked from Auto (pos = 3):

    acceleration, cylinders, displacement, horsepower, mpg, name,
    origin, weight, year

The following objects are masked from Auto (pos = 4):

    acceleration, cylinders, displacement, horsepower, mpg, name,
    origin, weight, year



[1] 18.79401

In [8]:
set.seed(2)


# Use the subset option and run a linear regression on the Auto data 
lm_fit <- lm(mpg ~ horsepower, data = Auto, subset = train)

# Estimate the response for all 392 observations and use the mean function
# to calculate the MSE of 196 observations in the validation set 
attach(Auto)

mean((mpg - predict(lm_fit, Auto))[-train]^2)

The following objects are masked from Auto (pos = 3):

    acceleration, cylinders, displacement, horsepower, mpg, name,
    origin, weight, year

The following objects are masked from Auto (pos = 4):

    acceleration, cylinders, displacement, horsepower, mpg, name,
    origin, weight, year

The following objects are masked from Auto (pos = 5):

    acceleration, cylinders, displacement, horsepower, mpg, name,
    origin, weight, year



[1] 23.26601

In [11]:
# We can use the poly() function to estimate the 
# test error for the quadratic regressions.
lm_fit2 <- lm(mpg ~ poly(horsepower, 2), data = Auto, subset = train)

attach(Auto)
mean((mpg - predict(lm_fit2, Auto))[-train]^2)

The following objects are masked from Auto (pos = 3):

    acceleration, cylinders, displacement, horsepower, mpg, name,
    origin, weight, year

The following objects are masked from Auto (pos = 4):

    acceleration, cylinders, displacement, horsepower, mpg, name,
    origin, weight, year

The following objects are masked from Auto (pos = 5):

    acceleration, cylinders, displacement, horsepower, mpg, name,
    origin, weight, year

The following objects are masked from Auto (pos = 6):

    acceleration, cylinders, displacement, horsepower, mpg, name,
    origin, weight, year

The following objects are masked from Auto (pos = 7):

    acceleration, cylinders, displacement, horsepower, mpg, name,
    origin, weight, year

The following objects are masked from Auto (pos = 8):

    acceleration, cylinders, displacement, horsepower, mpg, name,
    origin, weight, year



[1] 18.71646

In [12]:
# We can use the poly() function to estimate the 
# test error for the cubic regressions.
lm_fit3 <- lm(mpg ~ poly(horsepower, 3), data = Auto, subset = train)

attach(Auto)
mean((mpg - predict(lm_fit3, Auto))[-train]^2)

The following objects are masked from Auto (pos = 3):

    acceleration, cylinders, displacement, horsepower, mpg, name,
    origin, weight, year

The following objects are masked from Auto (pos = 4):

    acceleration, cylinders, displacement, horsepower, mpg, name,
    origin, weight, year

The following objects are masked from Auto (pos = 5):

    acceleration, cylinders, displacement, horsepower, mpg, name,
    origin, weight, year

The following objects are masked from Auto (pos = 6):

    acceleration, cylinders, displacement, horsepower, mpg, name,
    origin, weight, year

The following objects are masked from Auto (pos = 7):

    acceleration, cylinders, displacement, horsepower, mpg, name,
    origin, weight, year

The following objects are masked from Auto (pos = 8):

    acceleration, cylinders, displacement, horsepower, mpg, name,
    origin, weight, year

The following objects are masked from Auto (pos = 9):

    acceleration, cylinders, displacement, horsepower, mpg, nam

[1] 18.79401

## Leave-One-Out Cross-Validation

We will perform linear
regression using the glm() function rather than the lm() function because
the former can be used together with cv.glm(). The cv.glm() function is
part of the boot library.

In [13]:
library(boot)

glm_fit <- glm(mpg ~ horsepower, data = Auto)

cv_err <- cv.glm(Auto, glm_fit)

cv_err$delta

[1] 24.23151 24.23114

The cv.glm() function produces a list with several components. The two
numbers in the delta vector contain the cross-validation results. In this
case the numbers are identical (up to two decimal places) and correspond
to the LOOCV statistic.

Below, we discuss a situation in
which the two numbers differ. Our cross-validation estimate for the test
error is approximately 24.23.

We can repeat this procedure for increasingly complex polynomial fits. To automate the process, we use the for() function to initiate a for loop which iteratively fits polynomial regressions for polynomials of order i = 1to i = 10, computes the associated cross-validation error, and stores it in the ith element of the vector cv.error.

In [18]:
cv_error2 <- rep(0, 10)

for (i in 1:10) {
    glm_fit <- glm(mpg ~ poly(horsepower, i), data = Auto)
    cv_error2[i] <- cv.glm(Auto, glm_fit)$delta[1]
}

cv_error2

[1] 24.23151 19.24821 19.33498 19.42443 19.03321 18.97864 18.83305 18.96115
 [9] 19.06863 19.49093

## k-Fold Cross Validation

The cv.glm() function can also be used to implement k-fold CV. Below we
use k = 10, a common choice for k, on the Auto data set. We once again set
a random seed and initialize a vector in which we will store the CV errors
corresponding to the polynomial fits of orders one to ten.

In [16]:
set.seed(17)

cv_error10 <- rep(0, 10)

for (i in 1:10) {
    glm_fit <- glm(mpg ~ poly(horsepower, i), data = Auto)
    cv_error10[i] <- cv.glm(Auto, glm_fit, K = 10)$delta[1]
}

cv_error10

[1] 24.27207 19.26909 19.34805 19.29496 19.03198 18.89781 19.12061 19.14666
 [9] 18.87013 20.95520

## The Bootstrap

One of the great advantages of the bootstrap approach is that it can be
applied in almost all situations. No complicated mathematical calculations
are required. Performing a bootstrap analysis in R entails only two steps.
First, we must create a function that computes the statistic of interest.
Second, we use the boot() function, which is part of the boot library, to
perform the bootstrap by repeatedly sampling observations from the data
set with replacement.

The bootstrap approach can be used to assess the variability of the coefficient
estimates and predictions from a statistical learning method. Here
we use the bootstrap approach in order to assess the variability of the
estimates for β0 and β1, the intercept and slope terms for the linear regression
model that uses horsepower to predict mpg in the Auto data set.

In [19]:
boot_fn <- function(data, index) {

    coef(lm(mpg ~ horsepower, data = data, subset = index))

}

boot_fn(data = Auto, index = 1:392)

(Intercept)  horsepower 
 39.9358610  -0.1578447

The boot.fn() function can also be used in order to create bootstrap estimates
for the intercept and slope terms by randomly sampling from among
the observations with replacement. Here we give two examples.

In [21]:
set.seed(1)

boot_fn(data = Auto, sample(392, 392, replace = T))

(Intercept)  horsepower 
 40.3404517  -0.1634868

In [23]:
# Next, we use the boot() function to compute the standard errors of 1,000
# bootstrap estimates for the intercept and slope terms.

boot(Auto, boot_fn, 1000)


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = Auto, statistic = boot_fn, R = 1000)


Bootstrap Statistics :
      original        bias    std. error
t1* 39.9358610  0.0549915227 0.841925746
t2* -0.1578447 -0.0006210818 0.007348956

This indicates that the bootstrap estimate for SE( ˆ β0) is 0.84, and that the bootstrap estimate for SE( ˆ β1) is 0.0073. Standard formulas can be used to compute the standard errors for the regression coefficients in a linear model. These can be obtained using the summary() function.

In [24]:
summary(lm(mpg ~ horsepower, data = Auto))$coef

Estimate   Std. Error  t value   Pr(>|t|)     
(Intercept) 39.9358610 0.717498656  55.65984 1.220362e-187
horsepower  -0.1578447 0.006445501 -24.48914  7.031989e-81

The standard error estimates for ˆ β0 and ˆ β1 obtained using the formulas are 0.717 for the intercept and 0.0064 for the slope.
Interestingly, these are somewhat different from the estimates obtained using the bootstrap. Does this indicate a problem with the bootstrap? Infact, it suggests the opposite. The coefficients from the summary function rely on 2 assumptions: 
- they depend on the unknown parameter σ2, the noise variance.We then estimate σ2 using the RSS. Now although the formulas for the standard errors do not rely on the linear model being correct, the estimate for σ2 does.

- Secondly, the standard formulas assume (somewhat unrealistically) that the xi are fixed, and all the variability comes from the variation in the errors ϵi.


The bootstrap approach does not rely on any of these assumptions, and so it is likely giving a more accurate estimate of the standard errors of ˆ β0 and ˆ β1 than is the summary() function.

In [28]:
# Builing a finction for the Quadratic Model
boot_fn <- function(data, index) {
    coef(
        lm(mpg ~ horsepower + I(horsepower^2), data = data, subset = index)
    )
}

# Bootstrap estimates of the Coefficient Errors

set.seed(1)

boot(data = Auto, statistic = boot_fn, R = 1000)

# Standard Error estimates with linear regression
summary(lm(mpg ~ horsepower + I(horsepower^2), data = Auto))


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = Auto, statistic = boot_fn, R = 1000)


Bootstrap Statistics :
        original        bias     std. error
t1* 56.900099702  3.511640e-02 2.0300222526
t2* -0.466189630 -7.080834e-04 0.0324241984
t3*  0.001230536  2.840324e-06 0.0001172164


Call:
lm(formula = mpg ~ horsepower + I(horsepower^2), data = Auto)

Residuals:
     Min       1Q   Median       3Q      Max 
-14.7135  -2.5943  -0.0859   2.2868  15.8961 

Coefficients:
                  Estimate Std. Error t value Pr(>|t|)    
(Intercept)     56.9000997  1.8004268   31.60   <2e-16 ***
horsepower      -0.4661896  0.0311246  -14.98   <2e-16 ***
I(horsepower^2)  0.0012305  0.0001221   10.08   <2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 4.374 on 389 degrees of freedom
Multiple R-squared:  0.6876,	Adjusted R-squared:  0.686 
F-statistic:   428 on 2 and 389 DF,  p-value: < 2.2e-16
